exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java

```diff
@@ -69,29 +72,43 @@ public static PhysicalPlan getPlan(QueryContext context, String sql, Pointer<Str
     final SqlHandlerConfig config = new SqlHandlerConfig(context, parser);
 
     switch(sqlNode.getKind()){
-    case EXPLAIN:
-      handler = new ExplainHandler(config, textPlan);
-      break;
-    case SET_OPTION:
-      handler = new SetOptionHandler(context);
-      break;
-    case OTHER:
-      if(sqlNode instanceof SqlCreateTable) {
-        handler = ((DrillSqlCall)sqlNode).getSqlHandler(config, textPlan);
+      case EXPLAIN:
+        handler = new ExplainHandler(config, textPlan);
         break;
-      }
-
-      if (sqlNode instanceof DrillSqlCall) {
-        handler = ((DrillSqlCall)sqlNode).getSqlHandler(config);
+      case SET_OPTION:
+        handler = new SetOptionHandler(context);
         break;
-      }
-      // fallthrough
-    default:
-      handler = new DefaultSqlHandler(config, textPlan);
+      case OTHER:
+        if(sqlNode instanceof SqlCreateTable) {
+          handler = ((DrillSqlCall)sqlNode).getSqlHandler(config, textPlan);
+          break;
+        }
+
+        if (sqlNode instanceof DrillSqlCall) {
+          handler = ((DrillSqlCall)sqlNode).getSqlHandler(config);
+          break;
+        }
+        // fallthrough
+      default:
+        handler = new DefaultSqlHandler(config, textPlan);
     }
 
     try {
-      return handler.getPlan(sqlNode);
+      try {
+        return handler.getPlan(sqlNode);
+      } catch (UserException e) {
+        if (context.getOption(ExecConstants.DYNAMIC_UDF_SUPPORT_ENABLED).bool_val) {
+          final Throwable rootCause = ExceptionUtils.getRootCause(e);
+          if (rootCause instanceof SqlValidatorException
+              && StringUtils.contains(rootCause.getMessage(), "No match found for function signature")) {
+            if (context.getFunctionRegistry().loadRemoteFunctions()) {
```

Comment:
Rather than exposing the two-step logic here, can we wrap this in a single function? For example, reload( ) would do the two-part check. It would return true if something was reloaded (and a get plan can be retried), or false (meaning that nothing changed, no need to retry.)

Also, this is the root of the tree to consider for the concurrent load case: assume two identical queries hit this same Drillbit at the same time. Both flow down this same path. How do we ensure that both are retried (even though only the first one detects the need to refresh functions)? I suspect that this actually won't work.

Perhaps add the following. At the top of this function, do something like:

int ticket = context.getRegistryTicket( );

Then, in the failure case, do:

if ( context.reload( ticket ) ) ...

This says that if the registration changed (by us or any other thread) since we got the ticket, go ahead and retry.

The ticket itself is just a counter incremented each time new functions are loaded. reload( ) returns true if the ticket argument is less than the current value, or if new functions are available (which bumps the counter).

The ticket system seems like it should handle the concurrency issues. Please consider this or any other system that handles the concurren case correctly.

Finally, because this is so complex, we need good unit tests of this behavior; it can never be tested properly by bolting all of Drill together and testing from the outside by firing queries at Drill.